File: contest/ICPC/2018/欧洲/NWERC西北欧/data/fastestspeedrun/submissions/accepted/rgl-edmonds-lc.py
#!/usr/bin/env python
from heapq import heappush, heappop

class LinkCut:
  def __init__(self):
    self.p = self.l = self.r = None

  def root(s):
    return s.p is None or s.p.l is not s and s.p.r is not s

  def rotl(s):
    if not s.root():
      if s.p.l is s:
        s.p.l = s.l
      else:
        s.p.r = s.l
    s.l.p = s.p ; s.p = s.l
    s.l = s.p.r ; s.p.r = s
    if s.l is not None:
      s.l.p = s

  def rotr(s):
    if not s.root():
      if s.p.l is s:
        s.p.l = s.r
      else:
        s.p.r = s.r
    s.r.p = s.p ; s.p = s.r
    s.r = s.p.l ; s.p.l = s
    if s.r is not None:
      s.r.p = s

  def rotate(s):
    if s.p.l is s: s.p.rotl()
    else: s.p.rotr()

  def splay(s):
    while not s.root():
      if not s.p.root():
        if (s.p.l is s) == (s.p.p.l is s.p):
          s.p.rotate()
        else:
          s.rotate()
      s.rotate()

  def link(s,x):
    s.expose()
    x.splay()
    s.p = x ; x.l = s
    s.expose()

  def cut(s):
    s.splay()
    if s.r is not None:
      s.r.p = s.p
      s.r = None
    s.p = None

  def expose(s):
    i = s ; j = None
    while i is not None:
      i.splay()
      i.l = j ; j = i ; i = i.p
    s.splay()

  def findroot(s):
    s.expose()
    x = s
    while x.r is not None:
      x = x.r
    x.expose()
    return x

def minimum_cost_arborescence(edge):
  n = len(edge)

  pi = [[] for i in range(n)]
  for i in range(n):
    pi[i] = sorted([(edge[j][i],j) for j in range(n) if i != j])

  lc = [LinkCut() for i in range(n)]
  p = [-1 for i in range(n)]
  exists = [True for i in range(n)]

  def get_pi(x):
    while not exists[pi[x][0][1]] or edge[pi[x][0][1]][x] != pi[x][0][0]:
      heappop(pi[x])
    return pi[x][0][1]

  res = 0
  for i in range(1,n):
    if not exists[i]:
      continue

    while lc[get_pi(i)].findroot() is lc[i]:
      j = get_pi(i)

      old_pi = {i: edge[j][i]}
      merging = set([i])
      x = j
      while x != i:
        exists[x] = False
        old_pi[x] = edge[p[x]][x]
        merging.add(x)
        x = p[x]

      res += old_pi[i]

      cycle_children = []
      for a in range(1, i):
        if exists[a] and p[a] in merging:
          cycle_children += [a]
          res -= edge[p[a]][a]
          lc[a].cut()

      for a in range(n):
        if exists[a] and a != i:
          edge[a][i] -= old_pi[i]

      changed = set()
      for b in merging:
        if b != i:
          for a in range(n):
            if a != i and exists[a]:
              edge[a][i] = min(edge[a][i], edge[a][b] - old_pi[b])
              if edge[b][a] < edge[i][a]:
                edge[i][a] = min(edge[i][a], edge[b][a])
                changed.add(a)

      pi[i] = []
      for a in range(n):
        if exists[a] and a != i:
          if a in changed:
            heappush(pi[a], (edge[i][a], i))
          pi[i] += [(edge[a][i], a)]
      pi[i].sort()

      for a in cycle_children:
        res += edge[i][a]
        lc[a].link(lc[i])
        p[a] = i

    lc[i].link(lc[get_pi(i)])
    res += edge[get_pi(i)][i]
    p[i] = get_pi(i)

  return res

def main():
  n = int(raw_input())

  edge = [[2**62 for i in range(n+1)] for j in range(n+1)]
  for i in range(1,n+1):
    arr = map(int, raw_input().split())
    arr[2 + arr[0]] = arr[1]
    for j,w in enumerate(arr[2:]):
      edge[j][i] = w

  print minimum_cost_arborescence(edge)

if __name__ == '__main__':
  main()

File: contest/ICPC/2018/欧洲/NWERC西北欧/data/fastestspeedrun/submissions/accepted/rgl-edmonds-lc.cc
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

// https://en.wikipedia.org/wiki/Edmonds%27_algorithm
// With link-cut trees to improve performance from O(VE) to O(V^2logV)

struct LinkCut{/*{{{*/
  LinkCut *p,*l,*r;
  LinkCut(LinkCut* p=0): p(p), l(0), r(0) {}

  bool root() {return not p or p->l != this and p->r != this;}

  void rot(LinkCut* LinkCut::*a,LinkCut* LinkCut::*b) {
    if (not root()) (p->l==this?p->l:p->r)=(this->*a);
    (this->*a)->p=p, p=(this->*a);
    (this->*a)=(p->*b), (p->*b)=this;
    if (this->*a) (this->*a)->p=this;
  }

  void rotate() {
    if (p->l==this) p->rot(&LinkCut::l,&LinkCut::r);
    else            p->rot(&LinkCut::r,&LinkCut::l);
  }

  void splay() {
    while (not root()) {
      if (not p->root()) ((p->l==this)==(p->p->l==p)? p:this)->rotate();
      rotate();
    }
  }

  void link(LinkCut* x) {
    splay(); x->splay(); p=x; p->l=this; expose();
  }

  void cut() {
    splay(); if (r) r->p=p; p=r=0;
  }

  void expose() {
    for (LinkCut* i=this,*j=0; i; j=i,i=i->p) {
      i->splay();
      i->l=j;
    }
    splay();
  }

  LinkCut* findroot() {
    expose();
    LinkCut* x=this;
    while (x->r) x=x->r;
    x->expose();
    return x;
  }

};/*}}}*/

struct Graph{
  int const n;
  vector<vector<int32_t>> edge;

  Graph(int n): n(n) {edge.assign(n,vector<int32_t>(n,1<<30));}

  int64_t minimum_cost_arborescence(){
    vector<priority_queue<pair<int32_t,int>,vector<pair<int32_t,int>>,greater<pair<int32_t,int>>>> pi(n);
    for (int i=n; i-->1;)
      for (int j=n; j-->0;)
        if (i!=j) pi[i].emplace(edge[j][i],j);

    vector<LinkCut> lc(n);
    vector<bool> exists(n,true),merging(n,false);
    vector<int> par(n,-1);

    auto const get_pi=[&pi,&exists,this](int i){
      while (not exists[pi[i].top().second] or edge[pi[i].top().second][i]!=pi[i].top().first){
        pi[i].pop();
      }
      return pi[i].top().second;
    };

    int64_t res=0;
    for (int i=1; i<n; i++) if (exists[i]){
      int const j=get_pi(i);

      // If no conflict, make i a child of j.
      if (lc[j].findroot()!=&lc[i]){
        lc[i].link(&lc[j]);
        res+=edge[j][i];
        par[i]=j;
      }else{
        // Conflict! Rest of cycle goes through a common ancestor.
        vector<int> del={i};
        merging[i]=true;
        for (int x=j; x!=i; x=par[x]){
          del.push_back(x);
          merging[x]=true;
        }

        vector<int32_t> old_pi(n);
        for (int b: del){
          if (b==i){
            old_pi[b]=edge[j][b];
          }else{
            old_pi[b]=edge[par[b]][b];
            exists[b]=false;
          }
        }
        res+=edge[j][i];

        // Detach edges from the solution that went to deleted nodes.
        vector<int> cycle_children;
        for (int a=1; a<i; a++) if (exists[a] and merging[par[a]]) {
          cycle_children.push_back(a);
          res-=edge[par[a]][a];
          lc[a].cut();
        }

        // Merge deleted nodes' edges together, adding new edges for vertex i.
        for (int a=0; a<n; a++) if (exists[a] and a!=i){
          edge[a][i]-=old_pi[i];
        }
        for (int b: del) if (b!=i){
          for (int a=0; a<n; a++) if (a!=i and exists[a]){
            edge[a][i]=min(edge[a][i],edge[a][b]-old_pi[b]);
            edge[i][a]=min(edge[i][a],edge[b][a]);
          }
        }
        // Put the merged node back into the priority queues.
        for (int a=0; a<n; a++) if (exists[a] and a!=i){
          pi[a].emplace(edge[i][a],i);
          pi[i].emplace(edge[a][i],a);
        }
        // Reattach deleted solution edges to the merged vertex.
        for (int a: cycle_children){
          res+=edge[i][a];
          lc[a].link(&lc[i]);
          par[a]=i;
        }
        for (int b: del) {
          merging[b]=false;
        }
        --i;
      }
    }

    return res;
  }
};

int main(){
  ios::sync_with_stdio(false);
  int n; cin>>n;
  Graph g(n+1);

  for (int i=1; i<=n; i++){
    int x,t; cin>>x>>t;
    for (int j=0; j<=n; j++){
      cin>>g.edge[j][i];
    }
    if (x!=i){
      g.edge[x][i]=t;
    }
  }

  cout<<g.minimum_cost_arborescence()<<endl;
}
